src/main/resources/include/hail/Region.h

```diff
@@ -0,0 +1,121 @@
+#ifndef HAIL_REGION_H
+#define HAIL_REGION_H 1
+
+#include "hail/NativeObj.h"
+#include "hail/NativePtr.h"
+#include <cstdlib>
+#include <cstring>
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+namespace hail {
+
+class Region;
+using RegionPtr = std::shared_ptr<Region>;
+
+class Region : public NativeObj {
+private:
+  static constexpr ssize_t kChunkCap = 64*1024;
+  static constexpr ssize_t kMaxSmall = 4*1024;
+  static constexpr ssize_t kNumBigToKeep = 4;
+  
+  struct CharArray {
+    char buf_[1];
+  };
+  
+  class UniqueChunk {
+   public:
+    std::unique_ptr<CharArray> ptr_;
+    ssize_t size_;
+    
+    UniqueChunk() : ptr_(), size_(0) { }
+    UniqueChunk(ssize_t n) : ptr_(), size_(0) { alloc(n); }
+   
+    UniqueChunk(UniqueChunk&& b) = default;   
+    UniqueChunk& operator=(UniqueChunk&& b) = default;
+    
+    char* get() const { return ptr_.get()->buf_; }
+    
+    void alloc(ssize_t n) {
+      ptr_.reset((CharArray*)malloc(n));
+      size_ = n;
+    }
+    
+    bool empty() const { return !ptr_; }
+  };
+  
+public:
+  UniqueChunk chunk_;
```

Comment:
You mentioned being disappointed in the increase in lines of code. You could save some of the boilerplate of `UniqueChunk` and `CharArray` with
```C++
using UniqueChunk = std::unique_ptr<char,
                                    [](char* ptr){ free(ptr); }>
UniqueChunk make_chunk(ssize_t n) { return std::unique_ptr<char>((char*)malloc(n)); }
```
`unique_ptr` converts implicitly to `bool`, so `empty()` is covered. And instead of `chunk_.alloc(n);`, you can use `chunk_ = make_chunk(n);` This will move the new chunk into `chunk_`, because `make_chunk(n)` is an rvalue. That's probably one extra pointer copy, but compared to the cost of the `malloc` that's nothing.

Reply:
Since I'm using the same class for the big-allocations, it has to remember the size,
so unique_ptr on its own is not sufficient.